include/ccf/ds/pal.h

```diff
@@ -48,6 +49,47 @@ namespace ccf
     size_t peak_allocated_heap_size = 0;
   };
 
+  struct CpuidInfo
+  {
+    uint64_t eax;
+    uint64_t ebx;
+    uint64_t ecx;
+    uint64_t edx;
+  };
+
+  static void cpuid(CpuidInfo* info, uint64_t leaf, uint64_t subleaf)
+  {
+    asm volatile(
+      "cpuid"
+      : "=a"(info->eax), "=b"(info->ebx), "=c"(info->ecx), "=d"(info->edx)
+      : "a"(leaf), "c"(subleaf));
+  }
+
+  static bool is_intel_cpu()
```

Comment:
While the PAL is definitely the right place for this in principle, I'm not sure we want to expose all of this in our public API?

Reply:
I think that's true of pretty much everything in the `PAL` - it should move to private implementation at some point, but we don't want to shard the implementation.

Weighing this up with the _current_ PAL, I think we should give up on the class and expose these (already static, stateless) functions to a `pal` namespace. That way we can easily add implementations in several places (ie - some files only need `pal/attestation.h`), rather than trying to maintain a single monolothic PAL, and in turn move most of those implementations to private.

Reply:
Agreed. It's fine to go this route for now in the interest of time.